File: go/events/nats_event_bus.go
package events

import (
	"context"
	"encoding/json"
	"time"

	"github.com/nats-io/nats.go"
)

var _ Publisher = (*NatsEventBus)(nil)
var _ Subscriber = (*NatsEventBus)(nil)

type NatsEvent struct {
	Id        string      `json:"id"`
	EntityId  string      `json:"entity_id"`
	Topic     string      `json:"topic"`
	Payload   interface{} `json:"payload"`
	Timestamp time.Time   `json:"timestamp"`
}

type NatsEventBus struct {
	consumerGroup string
	conn          *nats.Conn
}

func NewNatsEventBus(
	consumerGroup string,
	conn *nats.Conn,
) *NatsEventBus {
	return &NatsEventBus{
		consumerGroup: consumerGroup,
		conn:          conn,
	}
}

func (eb *NatsEventBus) Publish(ctx context.Context, events ...*Event) error {
	for _, event := range events {
		natsEvent := NatsEvent{
			Id:        event.Id(),
			EntityId:  event.EntityId(),
			Topic:     event.Topic(),
			Payload:   event.Payload(),
			Timestamp: event.Timestamp(),
		}

		msg, err := json.Marshal(&natsEvent)
		if err != nil {
			return &ErrSerializingEvent{event, err}
		}

		if err := eb.conn.Publish(event.Topic(), msg); err != nil {
			return &ErrPublishingEvent{event, err}
		}
	}

	return nil
}

func (eb *NatsEventBus) Subscribe(ctx context.Context, subject string, handler Handler) error {
	_, err := eb.conn.QueueSubscribe(subject, eb.consumerGroup, func(msg *nats.Msg) {
		var natsEvent NatsEvent
		if err := json.Unmarshal(msg.Data, &natsEvent); err != nil {
			panic(&ErrDeserializingEvent{err})
		}

		event, err := NewEvent(
			natsEvent.Id,
			natsEvent.EntityId,
			natsEvent.Topic,
			natsEvent.Payload,
			natsEvent.Timestamp,
		)
		if err != nil {
			panic(err)
		}

		handler.Handle(ctx, event)
	})
	if err != nil {
		return &ErrSubscribingToSubject{subject, err}
	}

	return nil
}

File: typescript/src/events/nats-event-bus.ts
import { Codec, JSONCodec, NatsConnection, Subscription as NatsSubscription } from 'nats';
import { Buffer } from 'buffer';

import { Event, Handler, Publisher, Subscriber } from '../events';

export interface NatsEvent {
  id: string;
  entity_id: string;
  topic: string;
  payload: Uint8Array | string;
  timestamp: Date;
}

export class NatsEventBus implements Publisher, Subscriber {
  private consumerGroup: string;
  private conn: NatsConnection;
  private codec: Codec<NatsEvent>;

  constructor(consumerGroup: string, conn: NatsConnection) {
    this.consumerGroup = consumerGroup;
    this.conn = conn;
    this.codec = JSONCodec();
  }

  async publish(...events: Event[]): Promise<void> {
    for (const event of events) {
      const natsEvent: NatsEvent = {
        id: event.getId(),
        entity_id: event.getEntityId(),
        topic: event.getTopic(),
        payload: event.getPayload(),
        timestamp: event.getTimestamp(),
      };

      const msgJson = JSON.stringify(natsEvent);
      const msgBuff = Buffer.from(msgJson);

      this.conn.publish(event.getTopic(), new Uint8Array(msgBuff));
    }
  }

  async subscribe(subject: string, handler: Handler): Promise<void> {
    const sub = this.conn.subscribe(subject, {
      queue: this.consumerGroup,
    });

    (async (sub: NatsSubscription) => {
      for await (const msg of sub) {
        const natsEvent = this.codec.decode(msg.data);

        const event = new Event(
          natsEvent.id,
          natsEvent.entity_id,
          natsEvent.topic,
          typeof natsEvent.payload === 'string'
            ? new Uint8Array(Buffer.from(natsEvent.payload as string, 'base64'))
            : natsEvent.payload,
          natsEvent.timestamp,
        );

        await handler.handle(event);
      }
    })(sub);
  }
}
